spark/v3.4/spark-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergMergeInto.scala

```diff
@@ -0,0 +1,399 @@
+
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.Dataset
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.analysis.UnresolvedRelation
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.plans.logical.Assignment
+import org.apache.spark.sql.catalyst.plans.logical.DeleteAction
+import org.apache.spark.sql.catalyst.plans.logical.InsertAction
+import org.apache.spark.sql.catalyst.plans.logical.InsertStarAction
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.MergeAction
+import org.apache.spark.sql.catalyst.plans.logical.MergeIntoContext
+import org.apache.spark.sql.catalyst.plans.logical.UnresolvedMergeIntoIcebergTable
+import org.apache.spark.sql.catalyst.plans.logical.UpdateAction
+import org.apache.spark.sql.catalyst.plans.logical.UpdateStarAction
+import org.apache.spark.sql.functions.expr
+import scala.collection.JavaConverters
+
+
+object IcebergMergeInto {
+  /**
+   * Initialize an [[IcebergMergeIntoBuilder]].
+   *
+   * @param table : Target table name of the merge action
+   * @return [[IcebergMergeIntoBuilder]]
+   */
+  def apply(table: String): IcebergMergeIntoBuilder =
+    new IcebergMergeIntoBuilder(table, None, None, Seq.empty[MergeAction], Seq.empty[MergeAction])
+
+  /**
+   * Builder to specify an IcebergMergeInto action.
+   *
+   * It could be possible to provide any number of `whenMatched`
+   * and `whenNotMatched` actions.
+   *
+   *
+   * Scala Examples:
+   * {{{
+   *   val ds = ...
+   *   IcebergMergeInto("icebergTable")
+   *      .using(ds.as("source")
+   *      .when("source.id = icebergTable.id")
+   *      .whenMatched("source.op = U")
+   *      .updateAll()
+   *      .whenMatched("source.op = D)
+   *      .delete()
+   *      .whenNotMatched()
+   *      .insertAll()
+   * }}}
+   *
+   * JavaExamples:
+   * {{{
+   *   Dataset<Row> ds = ...
+   *   IcebergMergeInto.apply("icebergTable")
+   *      .using(ds.as("source")
+   *      .when("source.id = icebergTable.id")
+   *      .whenMatched("source.op = U")
+   *      .updateAll()
+   *      .whenMatched("source.op = D)
+   *      .delete()
+   *      .whenNotMatched()
+   *      .insertAll()
+   * }}}
+   *
+   */
+  class IcebergMergeIntoBuilder(
+                                 private val targetTable: String,
+                                 private val source: Option[Dataset[Row]],
+                                 private val whenCondition: Option[Expression],
+                                 private val whenMatchedActions: Seq[MergeAction],
+                                 private val whenNotMatchedActions: Seq[MergeAction]) {
+
+    /**
+     * Set the source dataset used during merge actions.
+     *
+     * @param source  : Dataset[Row]
+     * @return [[IcebergMergeIntoBuilder]]
+     */
+    def using(source: Dataset[Row]): IcebergMergeIntoBuilder =
```

Comment:
Why is the target table using the table name while the source table using a `Dataset`? Should we support the table name as well?

Reply:
@ConeyLiu In my opinion the main use of a Merge api is to handle the merge between a source dataset and a target table. To avoid having to register the dataset as a temporary table and perform the operation in sql. 

In case both the "delta" and the target table are already registered I don't see why use this api in favour of a simple SQL query.

Reply:
Hi @ConeyLiu can we consider this thread as solved?